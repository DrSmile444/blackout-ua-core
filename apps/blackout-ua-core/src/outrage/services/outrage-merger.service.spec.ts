import type { TestingModule } from '@nestjs/testing';
import { Test } from '@nestjs/testing';

import type { OutrageDto } from '@app/shared';
import {
  LightStatus,
  outrageMock3Origin,
  outrageMock4Change,
  outrageMock5Change2,
  OutrageParserService,
  OutrageRegion,
  OutrageType,
  stringToShift,
} from '@app/shared';

import { OutrageMergerService } from './outrage-merger.service';

const mockDate = new Date('2024-06-25');

describe('OutrageMergerService', () => {
  let outrageMergerService: OutrageMergerService;
  let outrageParserService: OutrageParserService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [OutrageMergerService, OutrageParserService],
    }).compile();

    outrageMergerService = module.get(OutrageMergerService);
    outrageParserService = module.get(OutrageParserService);
  });

  describe('parseTime', () => {
    it('should return time in minutes', () => {
      const result = outrageMergerService.parseTime('10:30');
      expect(result).toEqual(630);
    });
  });

  describe('mergeOutrages', () => {
    it('should merge 2 outrages', () => {
      const outrage1: OutrageDto = {
        date: mockDate,
        type: OutrageType.SCHEDULE,
        region: OutrageRegion.CHERKASY,
        shifts: [
          {
            start: stringToShift(new Date(), '10:00'),
            end: stringToShift(new Date(), '12:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '12:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const outrage2: OutrageDto = {
        date: mockDate,
        type: OutrageType.CHANGE,
        region: OutrageRegion.CHERKASY,
        changeCount: 1,
        shifts: [
          {
            start: stringToShift(new Date(), '12:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '18:00'),
            end: stringToShift(new Date(), '20:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const expected: OutrageDto = {
        date: mockDate,
        type: OutrageType.CHANGE,
        region: OutrageRegion.CHERKASY,
        changeCount: 1,
        shifts: [
          {
            start: stringToShift(new Date(), '10:00'),
            end: stringToShift(new Date(), '12:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '12:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '18:00'),
            end: stringToShift(new Date(), '20:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const actual = outrageMergerService.mergeOutrages([outrage1, outrage2]);
      expect(expected).toEqual(actual);
    });

    it('should merge 2 outrages with overlapping periods', () => {
      const outrage1: OutrageDto = {
        date: mockDate,
        type: OutrageType.SCHEDULE,
        region: OutrageRegion.CHERKASY,
        shifts: [
          {
            start: stringToShift(new Date(), '10:00'),
            end: stringToShift(new Date(), '12:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '12:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const outrage2: OutrageDto = {
        date: mockDate,
        type: OutrageType.CHANGE,
        region: OutrageRegion.CHERKASY,
        changeCount: 1,
        shifts: [
          {
            start: stringToShift(new Date(), '13:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '18:00'),
            end: stringToShift(new Date(), '20:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const expected: OutrageDto = {
        date: mockDate,
        type: OutrageType.CHANGE,
        region: OutrageRegion.CHERKASY,
        changeCount: 1,
        shifts: [
          {
            start: stringToShift(new Date(), '10:00'),
            end: stringToShift(new Date(), '12:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '12:00'),
            end: stringToShift(new Date(), '13:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '13:00'),
            end: stringToShift(new Date(), '14:00'),
            queues: [
              { queue: '1', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '2', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
          {
            start: stringToShift(new Date(), '18:00'),
            end: stringToShift(new Date(), '20:00'),
            queues: [
              { queue: '3', lightStatus: LightStatus.UNAVAILABLE },
              { queue: '4', lightStatus: LightStatus.UNAVAILABLE },
            ],
          },
        ],
      };

      const actual = outrageMergerService.mergeOutrages([outrage1, outrage2]);
      expect(expected).toEqual(actual);
    });

    it('should should merge real outrages', () => {
      const outrage = outrageParserService.parseMessage(mockDate, outrageMock3Origin, OutrageRegion.CHERKASY);
      const outrageChange = outrageParserService.parseMessage(mockDate, outrageMock4Change, OutrageRegion.CHERKASY);
      const outrageChange2 = outrageParserService.parseMessage(mockDate, outrageMock5Change2, OutrageRegion.CHERKASY);

      const mergedOutrage = outrageMergerService.mergeOutrages([outrage, outrageChange, outrageChange2]);

      expect(mergedOutrage).toMatchSnapshot();
    });
  });
});
