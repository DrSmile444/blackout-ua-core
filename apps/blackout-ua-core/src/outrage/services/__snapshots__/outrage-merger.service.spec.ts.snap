// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutrageMergerService mergeOutrages should should merge real outrages 1`] = `
{
  "date": 2024-06-23T21:00:00.000Z,
  "region": "cherkasy",
  "shifts": [
    {
      "end": 2024-06-23T22:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-23T21:00:00.000Z,
    },
    {
      "end": 2024-06-23T23:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-23T22:00:00.000Z,
    },
    {
      "end": 2024-06-24T00:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
      ],
      "start": 2024-06-23T23:00:00.000Z,
    },
    {
      "end": 2024-06-24T01:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
      ],
      "start": 2024-06-24T00:00:00.000Z,
    },
    {
      "end": 2024-06-24T02:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T01:00:00.000Z,
    },
    {
      "end": 2024-06-24T03:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T02:00:00.000Z,
    },
    {
      "end": 2024-06-24T04:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T03:00:00.000Z,
    },
    {
      "end": 2024-06-24T05:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T04:00:00.000Z,
    },
    {
      "end": 2024-06-24T06:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
      ],
      "start": 2024-06-24T05:00:00.000Z,
    },
    {
      "end": 2024-06-24T07:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
      ],
      "start": 2024-06-24T06:00:00.000Z,
    },
    {
      "end": 2024-06-24T08:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T07:00:00.000Z,
    },
    {
      "end": 2024-06-24T09:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T08:00:00.000Z,
    },
    {
      "end": 2024-06-24T10:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T09:00:00.000Z,
    },
    {
      "end": 2024-06-24T11:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
        {
          "lightStatus": 1,
          "queue": "5",
        },
      ],
      "start": 2024-06-24T10:00:00.000Z,
    },
    {
      "end": 2024-06-24T12:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T11:00:00.000Z,
    },
    {
      "end": 2024-06-24T13:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
        {
          "lightStatus": 1,
          "queue": "1",
        },
      ],
      "start": 2024-06-24T12:00:00.000Z,
    },
    {
      "end": 2024-06-24T14:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T13:00:00.000Z,
    },
    {
      "end": 2024-06-24T15:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
        {
          "lightStatus": 1,
          "queue": "3",
        },
      ],
      "start": 2024-06-24T14:00:00.000Z,
    },
    {
      "end": 2024-06-24T16:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
      ],
      "start": 2024-06-24T15:00:00.000Z,
    },
    {
      "end": 2024-06-24T17:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
        {
          "lightStatus": 1,
          "queue": "3",
        },
      ],
      "start": 2024-06-24T16:00:00.000Z,
    },
    {
      "end": 2024-06-24T18:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "3",
        },
        {
          "lightStatus": 1,
          "queue": "4",
        },
        {
          "lightStatus": 1,
          "queue": "5",
        },
      ],
      "start": 2024-06-24T17:00:00.000Z,
    },
    {
      "end": 2024-06-24T19:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "4",
        },
        {
          "lightStatus": 1,
          "queue": "5",
        },
        {
          "lightStatus": 1,
          "queue": "6",
        },
      ],
      "start": 2024-06-24T18:00:00.000Z,
    },
    {
      "end": 2024-06-24T20:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "6",
        },
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T19:00:00.000Z,
    },
    {
      "end": 2024-06-24T21:00:00.000Z,
      "queues": [
        {
          "lightStatus": 1,
          "queue": "1",
        },
        {
          "lightStatus": 1,
          "queue": "2",
        },
      ],
      "start": 2024-06-24T20:00:00.000Z,
    },
  ],
  "type": "change",
}
`;
